File: android/src/main/java/com/torusresearch/torusdirect/TorusDirectPlugin.java
package com.torusresearch.torusdirect;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.content.Context;
import androidx.annotation.NonNull;

import java.util.HashMap;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;

import org.torusresearch.torusdirect.TorusDirectSdk;
import org.torusresearch.torusdirect.interfaces.ILoginHandler;
import org.torusresearch.torusdirect.types.Auth0ClientOptions;
import org.torusresearch.torusdirect.types.DirectSdkArgs;
import org.torusresearch.torusdirect.types.LoginType;
import org.torusresearch.torusdirect.types.SubVerifierDetails;
import org.torusresearch.torusdirect.types.TorusLoginResponse;
import org.torusresearch.torusdirect.types.TorusNetwork;
import org.torusresearch.torusdirect.types.TorusVerifierUnionResponse;

import java.util.concurrent.ForkJoinPool;



/** TorusDirect */
public class TorusDirectPlugin  implements FlutterPlugin, MethodCallHandler, ActivityAware  {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private Context context;
  private Activity activity;
  private MethodChannel channel;
  private TorusDirectSdk torusDirectSDK;
  private SubVerifierDetails subVerifierDetails;


  public void  onDetachedFromActivity() {
    System.out.println("onDetachedFromActivity called");
  }

  public void  onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
    System.out.println("onReattachedToActivityForConfigChanges called");
  }

  public void onAttachedToActivity(ActivityPluginBinding binding) {
    activity = binding.getActivity();
  }

  public void onDetachedFromActivityForConfigChanges() {
    System.out.println("onDetachedFromActivityForConfigChanges called");
  }


  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    System.out.println("onAttachedToEngine called");
    channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "torus.flutter.dev/torus-direct");
    this.context = flutterPluginBinding.getApplicationContext();
    channel.setMethodCallHandler(this);
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    System.out.println("registerWith called");
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "torus.flutter.dev/torus-direct");
    channel.setMethodCallHandler(new TorusDirectPlugin());
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      
      case "setVerifierDetails":
          System.out.println(call.arguments);
        HashMap<String, String> args = (HashMap<String, String> ) call.arguments;
          String loginProviderString =  args.get("loginProvider");
          String clientId   =  args.get("clientId");
          String verifierName   = args.get("verifierName");

          this.subVerifierDetails = new SubVerifierDetails(
                  LoginType.valueOf(loginProviderString.toUpperCase()),verifierName,clientId, new Auth0ClientOptions.Auth0ClientOptionsBuilder("").build(),true);
        DirectSdkArgs directSdkArgs = new DirectSdkArgs("torusapp://io.flutter.app.FlutterApplication/redirect", TorusNetwork.TESTNET, "0x4023d2a0D330bF11426B12C6144Cfb96B7fa6183");
          this.torusDirectSDK  = new TorusDirectSdk(directSdkArgs, this.context);
        result.success(true);
        break;

      case "triggerLogin" :
        HashMap<String, String> torusLoginInfoMap = new HashMap();
        ForkJoinPool.commonPool().submit(() -> {
            TorusLoginResponse torusLoginResponse = this.torusDirectSDK.triggerLogin(subVerifierDetails).join();
          TorusVerifierUnionResponse userInfo = torusLoginResponse.getUserInfo();
          torusLoginInfoMap.put("privateKey", torusLoginResponse.getPrivateKey());
          torusLoginInfoMap.put("publicAddress", torusLoginResponse.getPublicAddress());
          torusLoginInfoMap.put("email",userInfo.getEmail());
          torusLoginInfoMap.put("name",userInfo.getName());
          torusLoginInfoMap.put("id",userInfo.getVerifierId());
          torusLoginInfoMap.put("profileImage",userInfo.getProfileImage());

          new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
              result.success(torusLoginInfoMap);
            }
          });
        });
        break;  
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}

File: android/src/main/kotlin/org/torusresearch/flutter/torus_direct/TorusDirectPlugin.kt
package org.torusresearch.flutter.torus_direct

import android.app.Activity
import android.content.Context
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.torusresearch.torusdirect.TorusDirectSdk
import org.torusresearch.torusdirect.types.*
import org.torusresearch.torusdirect.utils.Helpers.unwrapCompletionException

/** TorusDirectPlugin */
class TorusDirectPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    private var activity: Activity? = null
    private lateinit var context: Context
    private lateinit var torusDirectArgs: DirectSdkArgs

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "torus_direct")
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        CoroutineScope(Dispatchers.Default).launch {
            try {
                val response = runMethodCall(call)
                launch(Dispatchers.Main) { result.success(response) }
            } catch (e: NotImplementedError) {
                launch(Dispatchers.Main) { result.notImplemented() }
            } catch (e: Throwable) {
                launch(Dispatchers.Main) {
                    val unwrappedError = unwrapCompletionException(e)
                    result.error(unwrappedError::class.simpleName, unwrappedError.message, null)
                }
            }
        }
    }

    private fun runMethodCall(@NonNull call: MethodCall): Any? {
        when (call.method) {
            "init" -> {
                torusDirectArgs = DirectSdkArgs(
                    call.argument("browserRedirectUri"),
                    TorusNetwork.valueOfLabel(call.argument("network")),
                    call.argument("redirectUri")
                )
                Log.d(
                    "${TorusDirectPlugin::class.qualifiedName}#init",
                    "network=${torusDirectArgs.network}, redirectUri=${torusDirectArgs.redirectUri}"
                )
                return null
            }
            "triggerLogin" -> {
                val torusDirectSdk =
                    TorusDirectSdk(torusDirectArgs, activity ?: context)
                val torusResponse = torusDirectSdk.triggerLogin(
                    SubVerifierDetails(
                        LoginType.valueOfLabel(call.argument("typeOfLogin")),
                        call.argument<String>("verifier"),
                        call.argument<String>("clientId"),
                        mapJwtParams(call.argument("jwtParams")),
                        activity == null
                    )
                ).join()
                Log.d(
                    "${TorusDirectPlugin::class.qualifiedName}#triggerLogin",
                    "publicAddress=${torusResponse.publicAddress}"
                )
                return mapOf(
                    "userInfo" to torusResponse.userInfo,
                    "publicAddress" to torusResponse.publicAddress,
                    "privateKey" to torusResponse.privateKey
                )
            }
            "getTorusKey" -> {
                val torusDirectSdk =
                    TorusDirectSdk(torusDirectArgs, activity ?: context)
                val torusResponse = torusDirectSdk.getTorusKey(
                    call.argument("verifier"),
                    call.argument("verifierId"),
                    call.argument("verifierParams"),
                    call.argument("idToken")
                ).join()
                Log.d(
                    "${TorusDirectPlugin::class.qualifiedName}#getTorusKey",
                    "publicAddress=${torusResponse.publicAddress}"
                )
                return mapOf(
                    "publicAddress" to torusResponse.publicAddress,
                    "privateKey" to torusResponse.privateKey
                )
            }
        }
        throw NotImplementedError()
    }

    private fun mapJwtParams(jwtParams: Map<String, Any>?): Auth0ClientOptions? {
        if (jwtParams == null || !jwtParams.containsKey("domain")) {
            return Auth0ClientOptions.Auth0ClientOptionsBuilder("").build()
        }

        val builder = Auth0ClientOptions.Auth0ClientOptionsBuilder(jwtParams["domain"] as String)
        if (jwtParams.containsKey("isVerifierIdCaseSensitive")) {
            builder.setVerifierIdCaseSensitive(jwtParams["isVerifierIdCaseSensitive"] as Boolean)
        }
        if (jwtParams.containsKey("client_id")) {
            builder.setClient_id(jwtParams["client_id"] as String)
        }
        if (jwtParams.containsKey("leeway")) {
            builder.setLeeway(jwtParams["leeway"] as String)
        }
        if (jwtParams.containsKey("verifierIdField")) {
            builder.setVerifierIdField(jwtParams["verifierIdField"] as String)
        }
        if (jwtParams.containsKey("display")) {
            builder.setDisplay(Display.valueOfLabel(jwtParams["display"] as String))
        }
        if (jwtParams.containsKey("prompt")) {
            builder.setPrompt(Prompt.valueOfLabel(jwtParams["prompt"] as String))
        }
        if (jwtParams.containsKey("max_age")) {
            builder.setMax_age(jwtParams["max_age"] as String)
        }
        if (jwtParams.containsKey("ui_locales")) {
            builder.setUi_locales(jwtParams["ui_locales"] as String)
        }
        if (jwtParams.containsKey("id_token_hint")) {
            builder.setId_token_hint(jwtParams["id_token_hint"] as String)
        }
        if (jwtParams.containsKey("login_hint")) {
            builder.setLogin_hint(jwtParams["login_hint"] as String)
        }
        if (jwtParams.containsKey("acr_values")) {
            builder.setAcr_values(jwtParams["acr_values"] as String)
        }
        if (jwtParams.containsKey("scope")) {
            builder.setScope(jwtParams["scope"] as String)
        }
        if (jwtParams.containsKey("audience")) {
            builder.setAudience(jwtParams["audience"] as String)
        }
        if (jwtParams.containsKey("connection")) {
            builder.setConnection(jwtParams["connection"] as String)
        }
        if (jwtParams.containsKey("additionalParams")) {
            builder.setAdditionalParams(jwtParams["additionalParams"] as HashMap<String, String>)
        }

        return builder.build()
    }
}
